File: frontend/src/lib/humanScore.js
/**
 * Human Score report logic: what the AI "said" per dimension and how to render
 * the report as Markdown. Kept out of the form component so it can be unit-tested
 * and reused. Modelled on data/human/human_score_W28.md.
 */
import { HUMAN_DIMENSIONS, EVIDENCE_SOURCES } from './constants'
import { prepareAgentCard } from './agentDisplay'
import { classifyBias, summarizeModelAgreement } from './bias'

export const PLACEHOLDER = '________'

const BREAKDOWN_LABELS = {
  macro: 'Macro',
  technical: 'Technical',
  almanac: 'Almanac',
  aiAgreement: 'AI Agreement',
  wildCard: 'Wild Card',
}

/** Per-dimension summary of the AI outputs, shown in the "AI Said" column. */
export function aiSaidFor(outputs = {}) {
  const almanac = prepareAgentCard('almanac', outputs.almanac)
  const macro = prepareAgentCard('macro', outputs.macro)
  const technical = prepareAgentCard('technical', outputs.technical)

  return {
    macro: macro?.bias || '—',
    technical: technical?.bias || '—',
    almanac: almanac?.bias || '—',
    aiAgreement: summarizeModelAgreement(outputs.llmComparison),
    wildCard: 'nothing specifically flagged',
  }
}

function signed(score) {
  const n = Number(score) || 0
  return n > 0 ? `+${n}` : `${n}`
}

/** Format a team score with an explicit plus sign for positives. */
export function formatSignedScore(score) {
  return signed(score)
}

/** "2026-W28" / "W28" → "Week 28" for the report title. */
export function weekTitleLabel(week) {
  const m = String(week || '').match(/W(\d{1,2})/i)
  return m ? `Week ${Number(m[1])}` : week || '—'
}

/** Build the bold AI Consensus line, matching W28 style when model counts are known. */
export function formatConsensusHeading(consensus, llm) {
  const base = (consensus || '—').trim()
  const models = Array.isArray(llm?.models) ? llm.models : []
  if (!models.length || /\(\d+\s+of\s+\d+/i.test(base)) {
    return `**${base}**`
  }
  const direction = classifyBias(llm?.finalConsensus || base)
  const matching = models.filter(m => classifyBias(m?.consensus) === direction).length
  return `**${base} (${matching} of ${models.length} models)**`
}

function scoreBreakdown(form) {
  return HUMAN_DIMENSIONS.map(d => {
    const n = Number(form?.scores?.[d.key]) || 0
    return `${BREAKDOWN_LABELS[d.key]} ${signed(n)}`
  }).join(' + ')
}

/** Render a submitted report as Markdown (Copy as Markdown / export). */
export function buildHumanScoreMarkdown(form, ctx) {
  const {
    week = '—',
    consensus = '—',
    aiSaid = {},
    total = 0,
    llmComparison = null,
  } = ctx || {}
  const lines = []
  const titleWeek = weekTitleLabel(week)

  lines.push(`# Human Score Analyst Output — ${titleWeek}`, '')
  lines.push('## AI Consensus', '')
  lines.push(formatConsensusHeading(consensus, llmComparison), '')
  lines.push('---', '')

  lines.push('## Human Score Table', '')
  lines.push(
    '| Dimension                         | AI Said                                   | Team Score | Team Reasoning |',
  )
  lines.push(
    '| --------------------------------- | ----------------------------------------- | :--------: | -------------- |',
  )
  HUMAN_DIMENSIONS.forEach(d => {
    const score = form.scores?.[d.key]
    const reason = (form.reasoning?.[d.key] || PLACEHOLDER).replace(/\|/g, '/')
    const said = String(aiSaid[d.key] ?? '—').replace(/\|/g, '/')
    lines.push(
      `| **${d.label}** | ${said} | **${signed(score)}** | ${reason} |`,
    )
  })
  lines.push('', '---', '')

  lines.push('## Human Score Total', '')
  lines.push(`**${signed(total)}**`, '')
  lines.push(`(${scoreBreakdown(form)})`, '')
  lines.push('---', '')

  lines.push('## Five-Dimension Judgement', '')
  HUMAN_DIMENSIONS.forEach((d, i) => {
    const score = form.scores?.[d.key]
    const body = form.reasoning?.[d.key] || PLACEHOLDER
    lines.push(`### ${i + 1}. ${d.label} — Score: ${signed(score)}`, '')
    lines.push(body, '')
  })
  lines.push('---', '')

  lines.push('## Human Call', '', `**${form.humanCall || 'Neutral'}**`, '', '---', '')
  lines.push('## Confidence', '', `**${form.confidence || 'Medium'}**`, '', '---', '')
  lines.push('## Override Paragraph', '', form.overrideParagraph || PLACEHOLDER, '', '---', '')
  lines.push('## Wild Card Insight', '', form.wildCardInsight || PLACEHOLDER, '', '---', '')
  lines.push('## Invalidation Condition', '', form.invalidation || PLACEHOLDER, '', '---', '')
  lines.push('## Evidence Used', '')
  EVIDENCE_SOURCES.filter(s => form.evidence?.[s.key]).forEach(s => lines.push(`* ${s.label}`))

  return lines.join('\n')
}

/** Sum of the five team dimension scores. */
export function humanScoreTotal(form) {
  return HUMAN_DIMENSIONS.reduce((sum, d) => sum + (Number(form?.scores?.[d.key]) || 0), 0)
}

/** Bundle form + context for display and export. */
export function buildHumanScoreReport(form, { week, outputs, predictionDate }) {
  if (!form) return null
  const aiSaid = aiSaidFor(outputs)
  const consensus = outputs?.llmComparison?.finalConsensus || 'Pending — run the pipeline'
  return {
    form,
    week,
    predictionDate,
    consensus,
    aiSaid,
    total: humanScoreTotal(form),
    llmComparison: outputs?.llmComparison || null,
  }
}

File: backend/server/human_score.py
"""Human score persist: JSON source of truth, MD rendered for display/archive."""

from __future__ import annotations

import json
import re
from pathlib import Path

from agents.paths import DATA_DIR
from server.files import write_markdown

_DIMS = (
    ("macro", "Macro / News Weight", "Macro"),
    ("technical", "Technical Structure", "Technical"),
    ("almanac", "Almanac Seasonal Weight", "Almanac"),
    ("aiAgreement", "AI Model Agreement Quality", "AI Agreement"),
    ("wildCard", "Wild Card / Human Observation", "Wild Card"),
)
_EVIDENCE = (
    ("almanac", "R3 Almanac Agent Output"),
    ("macro", "R4 Macro Agent Output"),
    ("technical", "R5 Technical Agent Output"),
    ("llm", "R6 LLM Comparison Output"),
)
_PLACEHOLDER = "________"


def _signed(score) -> str:
    try:
        n = int(score)
    except (TypeError, ValueError):
        n = 0
    return f"+{n}" if n > 0 else str(n)


def _total(scores: dict) -> int:
    return sum(int(scores.get(k) or 0) for k, _, _ in _DIMS)


def render_human_score_markdown(report: dict) -> str:
    """W28-style markdown from a report JSON payload."""
    form = report.get("form") or {}
    scores = form.get("scores") or {}
    reasoning = form.get("reasoning") or {}
    evidence = form.get("evidence") or {}
    ai_said = report.get("aiSaid") or {}
    week = str(report.get("week") or "—")
    m = re.search(r"W(\d{1,2})", week, re.I)
    title = f"Week {int(m.group(1))}" if m else week
    total = report.get("total")
    if total is None:
        total = _total(scores)
    consensus = (report.get("consensus") or "—").strip()

    table = [
        "| Dimension                         | AI Said                                   | Team Score | Team Reasoning |",
        "| --------------------------------- | ----------------------------------------- | :--------: | -------------- |",
    ]
    for key, label, _ in _DIMS:
        said = str(ai_said.get(key) or "—").replace("|", "/")
        reason = str(reasoning.get(key) or _PLACEHOLDER).replace("|", "/")
        table.append(f"| **{label}** | {said} | **{_signed(scores.get(key))}** | {reason} |")

    judgements = []
    for i, (key, label, _) in enumerate(_DIMS, 1):
        judgements.append(
            f"### {i}. {label} — Score: {_signed(scores.get(key))}\n\n"
            f"{reasoning.get(key) or _PLACEHOLDER}\n"
        )

    breakdown = " + ".join(f"{short} {_signed(scores.get(key))}" for key, _, short in _DIMS)
    evidence_lines = [f"* {label}" for key, label in _EVIDENCE if evidence.get(key)]

    parts = [
        f"# Human Score Analyst Output — {title}",
        "",
        "## AI Consensus",
        "",
        f"**{consensus}**",
        "",
        "---",
        "",
        "## Human Score Table",
        "",
        *table,
        "",
        "---",
        "",
        "## Human Score Total",
        "",
        f"**{_signed(total)}**",
        "",
        f"({breakdown})",
        "",
        "---",
        "",
        "## Five-Dimension Judgement",
        "",
        *judgements,
        "---",
        "",
        "## Human Call",
        "",
        f"**{form.get('humanCall') or 'Neutral'}**",
        "",
        "---",
        "",
        "## Confidence",
        "",
        f"**{form.get('confidence') or 'Medium'}**",
        "",
        "---",
        "",
        "## Override Paragraph",
        "",
        form.get("overrideParagraph") or _PLACEHOLDER,
        "",
        "---",
        "",
        "## Wild Card Insight",
        "",
        form.get("wildCardInsight") or _PLACEHOLDER,
        "",
        "---",
        "",
        "## Invalidation Condition",
        "",
        form.get("invalidation") or _PLACEHOLDER,
        "",
        "---",
        "",
        "## Evidence Used",
        "",
        *evidence_lines,
    ]
    return "\n".join(parts)


def save_human_score(stem: str, report: dict, *, data_root: Path | None = None) -> Path:
    """Write human_score_{stem}.json + rendered .md. Overwrites if present."""
    stem = stem.upper()
    if not re.fullmatch(r"W\d{2}", stem):
        raise ValueError(f"Invalid week stem: {stem!r}")
    if not isinstance(report, dict) or not isinstance(report.get("form"), dict):
        raise ValueError("Human score report requires a form object")

    root = data_root or DATA_DIR
    payload = dict(report)
    payload.setdefault("week", stem)
    scores = (payload.get("form") or {}).get("scores") or {}
    if payload.get("total") is None:
        payload["total"] = _total(scores)

    human_dir = root / "human"
    human_dir.mkdir(parents=True, exist_ok=True)
    (human_dir / f"human_score_{stem}.json").write_text(
        json.dumps(payload, indent=2) + "\n", encoding="utf-8"
    )
    return write_markdown(
        human_dir / f"human_score_{stem}.md",
        render_human_score_markdown(payload),
    )


def load_human_score_json(stem: str, *, data_root: Path | None = None) -> dict | None:
    """Raw report dict from human_score_{stem}.json, or None."""
    path = (data_root or DATA_DIR) / "human" / f"human_score_{stem}.json"
    if not path.exists():
        return None
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    if not isinstance(data, dict) or not isinstance(data.get("form"), dict):
        return None
    return data
